services/cd-service/pkg/repository/transformer.go

```diff
@@ -2675,12 +2677,60 @@ func (c *ReleaseTrain) Transform(
 		if state.MaxNumThreads > 0 {
 			releaseTrainErrGroup.SetLimit(state.MaxNumThreads)
 		}
-		for _, envName := range envNames {
-			trainGroup := conversion.FromString(targetGroupName)
-			envNameLocal := envName
-			releaseTrainErrGroup.Go(func() error {
-				return c.runEnvReleaseTrainBackground(ctx, state, t, envNameLocal, trainGroup, envGroupConfigs, configs, allLatestReleases)
-			})
+		if state.ParallelismOneTransaction {
+			type ChannelData struct {
+				prognosis *ReleaseTrainEnvironmentPrognosis
+				train     *envReleaseTrain
+			}
+			var prognosisResultChannel = make(chan *ChannelData, 42)
+			expectedNumPrognoses := len(envNames)
+			for _, envName := range envNames {
+				trainGroup := conversion.FromString(targetGroupName)
+				envNameLocal := envName
+				releaseTrainErrGroup.Go(func() error {
+					span, ctx, onErr := tracing.StartSpanFromContext(ctx, "EnvReleaseTrain Transform")
+					defer span.Finish()
+
+					train := &envReleaseTrain{
+						Parent:                 c,
+						Env:                    envName,
+						EnvConfigs:             configs,
+						EnvGroupConfigs:        envGroupConfigs,
+						WriteCommitData:        c.WriteCommitData,
+						TrainGroup:             trainGroup,
+						TransformerEslVersion:  c.TransformerEslVersion,
+						CiLink:                 c.CiLink,
+						AllLatestReleasesCache: allLatestReleases,
+					}
+
+					prognosis, err := train.runEnvPrognosisBackground(ctx, state, t, envNameLocal, trainGroup, envGroupConfigs, configs, allLatestReleases)
+					if err != nil {
+						return onErr(err)
+					}
+					prognosisResultChannel <- &ChannelData{
+						prognosis: prognosis,
+						train:     train,
+					}
+					return nil
+				})
+			}
+			for i := 0; i < expectedNumPrognoses; i++ {
+				// waiting for prognoses to be done
+				var result = <-prognosisResultChannel
```

Comment:
Since this channel is buffered, the code here will not wait for all of the routines to finish. It just continues with any result that it gets. I think we can use something like 
```
releaseTrainErrGroup.Wait()
```
Before the for line.

Reply:
I solved it in a different way.
It's actually ok that it "continues with any result" because it loops over all results and the order does not matter.
I changed it now, so that the go routines never return an error, they just write the error to the channel.
This way, we do not have to wait for `.Wait()`, because it is already clear that the routines are done from the content of the channel.